js/binance.js

```diff
@@ -784,6 +785,15 @@ module.exports = class binance extends Exchange {
         else
             request['orderId'] = parseInt (id);
         let response = await this.privateGetOrder (this.extend (request, params));
+        if(response['fills'] === undefined && this.options['loadTradesForOrder']) {
+            let requestTrades = {
+                'symbol': market['id'],
+                'startTime': request['transactTime'],
+            };
+            let trades = await this.privateGetMyTrades (requestTrades); // temporary slow workaround todo implement caching layer
```

Comment:
Sorry, but this violates the design rules of the lib: one http request per unified method call. We only do more than one call if that is absolutely necessary from the library perspective. So, if the user wants order trades, instead of patching the `fetchOrder` method in the lib we say the user should call `fetchOrder` and then call `fetchMyTrades` to get the fills. In other words, if it requires two calls – it should live in the userland outside of the lib.

Reply:
Hm. Wow now I'm confused and concerned again... 
1) Where did you write down all these design rules? I admit I didnt read all docs.
2) Isnt the goal to unify the usage of exchanges (same function calls, same returned data structures for all exchanges)? 

How can you achieve a unified object/API with one single call for all exchanges if the exchange returns uncomplete data? Btw. you are doing two calls (fetching always market data) which is violating this design rule already(?). Sounds like you want ccxt to just be a single call REST wrapper instead of becoming a unifying API for multiple exchanges. Doesnt make sense for me. **Please help me understanding the idea/design behind ccxt (and please share the current road map).** How do you plan to solve these issues with ccxt that developers need to handle and cache each exchange individually? Sometimes it's necessary to revise made decisions to bring a project forward. If ccxt just wants to stay a REST wrapper, I could also use the REST APIs of the exchanges because I need to deal anyway with missing data (after extensive testing of each exchange). Then there is unfortunately absolutely no real reason to use (and contribute) to ccxt if I need to handle each exchange individually anyway. :-\

Reply:
I can't say that those are written rules, i mostly mean the generic understanding of library design in the first place – if you're going to design a lib, then you'd want it to remain functional and flexible – in that case you don't want to decide for the user whether two calls are good for the user or not. You should let the user decide. In your implementation you do that second call optionally, but there's still tons of errors introduced in that implementation and those errors are caused by the attempt to wrap the "unwrappable" thing, let me explain. When you do `privateGetMyTrades` your implementation will return just the most recent trades and because you're doing that on the client side for an old order it will either return no fills (empty array) or will return just the partial fill info (which is worse – some fills might not be there). So, if I'm calling your implementation of `fetchOrder` for one of my old orders – your implementation kinda does two calls for me (which I never asked for) and it will return an empty or a partial array, making the second call redundant in this case. Also, because a limit order can be filled partially, those fills don't have to correspond to the transactTime inside the order. So the user will have to traverse his trades with a subsequent calls to `fetchMyTrades` anyway to actually get his fill history. In other words, that immediate second call doesn't work for algotraders who do thousands of limit orders per day (i'm talking of the situation when you enable that option and don't get those fills). If you want to do that properly, you will have to build a caching and data-tracking subsystem to collect users trades periodically, but that will require a data-abstraction layer, which, we think, should be built on top of this library, but not into this library, because it often depends on your user-specific goals. 

We suggest another approach that we think is better and more flexible – instead of doing that second call for the user – give the user a convenient tool (more methods in the toolset) to do the call *when it is appropriate for the user* (don't decide for the user). So, the library never guarantees that you will get the *full order info* from `fetchOrder` – that isn't the goal of this library. The goal of this library is to guarantee that you will be able to get **as much unified info as possible to get from the exchange API in one call/http request** without adding a lot of caching overhead and with as less memusage as possible (quicky). So, if the exchange in question does not provide the fills in `fetchOrder` endpoint – we won't call that for the user, but we will provide a unified API call to get the fills either separately (`fetchMyTrades`) or for a specific order (`fetchOrderTrades`) – that is, again, if the exchange in question provides that. In this scenario, the user *should never* rely on `fills` within the order, instead the user should be aware (from the Manual) that `fills` might be missing in the order, thus, the user has to code for the logic that will fetch those fills – and for that we provide some unified methods as well – having the set of all those unified methods covers most of exchanges. So, yes, the goal of this library is to unify as much as possible while remaining a thin wrapper.

If you look at the most utilitarian and the most flexible and therefore the most widely used and successful libraries many of them stick to this approach – instead of doing two calls for the user in a complex wrapper – they will split the complex call into two smaller pieces and will provide the interface to the two calls separately so that the user calls any of them **when the user wants** (not when *the library wants* and not when *the author of the library wants* and not when *the contributors to the library want*). 

We never decide for the user – but we provide the ways for the user to decide. Hope it answers your questions.

Reply:
> You see its quite complicated to use ccxt. And this needs to be tested and done for EACH exchange someone wants to use as you can be never sure what the exchange is doing (as ccxt is not protecting an app developer from this exchange mess). AND you need to test it for each new ccxt release as some releases are breaking the functionality (or silently changing something). This last issue and the missing unifying can make it easier and safer to implement an exchange without ccxt. Unfortunately.

Let me clarify with a simple code snippet, to show that we're actually talking of the same thing, but we're just looking at it from different angles.

Think as if you're designing a library not for yourself, but for a generic user. You wouldn't want to add the second call just for Binance to `fetchOrder`. Instead you'd want to add `fetchOrderWithFills` that would do:

```JavaScript

// also NOT ideal, but a bit better than doing that inside fetchOrder...
// just for the purpose of explanation

async fetchOrderWithFills (id, symbol = undefined, params = {}) { // this does not have to be available in the library
    const order = await this.fetchOrder (id, symbol, params); // this has to be available in the library
    const since = order['lastTradeTimestamp'];
    const limit = undefined;
    const trades = await this.fetchMyTrades (symbol, since, limit, params) // this has to be available in the library
    // ...
    // match and filter...
    // ...
    order['fills'] = result;
    return order;
}
```

Does that snippet clarify the goals of this library? Because if you're designing the library from bottom upwards – you should agree to the comments in the snippet above.

The point is that you want it to implement all possible wrappers from the very beginning, but this is not how it works, unfortunately ) If you could make it all complete and ideal from the very start – it would be finished already ) So, we are doing a work in progress that has among one of its goals the goal of designing enough functionality to build the ultimate fetchOrderWrapper that you'd like to see there from the very start ) I hope this clarifies the difference i'm trying to explain )

Reply:
> When you do `privateGetMyTrades` your implementation will return a limited set of trades and because you're matching on the client side for an old order it will either return no fills (empty array) or will return just the partial fill info (which is worse – some fills might not be there).

Exactly and thats what I'm always complaining ;-) If ccxt would use its own database/cache (e.g. sqlite) there would be often no need for extra calls. And if I got all trades or not could be easily improved (for binance, this case) to check the filled amount vs the total amount in trades. 

> So, if I'm calling your implementation of fetchOrder for one of my old orders – your implementation kinda does two calls for me (which I never asked for) and it will return an empty or a partial array, making the second call redundant in this case. Also, because a limit order can be filled partially, those fills don't have to correspond to the transactTime inside the order. 

To fix the issue with unnecessary call (you are right, without fills there needs to be no call -> bug), I simply need to check for binance if there is a filling (then there is a trade). If the transactTime is not a good measurement we should simply remove it.

I'm aware of it but it's impossible to fix things properly will all your "not written and fixed design decisions" that are somehow fixed because you reject other code. And in case no trade 

>  In this scenario, the user should never rely on fills within the order, instead the user should be aware (from the Manual) that fills might be missing in the order, thus, the user has to code for the logic that will fetch those fills – and for that we provide some unified methods as well – having the set of all those unified methods covers most of exchanges. 

But then you need to mention that EACH value except of id is optional. Not just fills. And in case of binance also the fees are not present if the trades are not filled.

> We suggest another approach that we think is better and more flexible – instead of doing that second call for the user – give the user a convenient tool (more methods in the toolset) to do the call when it is appropriate for the user (don't decide for the user).

Then I would suggest to introduce a method `fetchOrderAndTradesUsingCache` to bring my requirements into a function name.

>  If you want to do that properly, you will have to build a caching and data-tracking subsystem to collect users trades periodically, but that will require a data-abstraction layer, which, we think, should be built on top of this library, but not into this library, because it often depends on your app-specific goals.

I dont think so. I'm doing this already but the fact that I need to implement it for each exchange individually to get all data brings up the question why I'm really using ccxt. I didnt expect it to be just a REST wrapper with some basic unifying (like symbols).

> If you look at the most utilitarian and the most flexible and therefore the most widely used and successful libraries many of them stick to this approach

Could you provide some examples please? So we can see in detail if this is comparable to this library. I really would like to bring this library success so I just wanna see it clearly.

> when the user wants 

To be honest: at the moment it looks to me its very centralized managed, and lots of developers would really appreciate if they dont need to deal with each exchange individually. I was wondering why not every commercial project starts using ccxt. Now I know why ;-) Breaking changes and no unifying doesnt make it very attractive from business perspective.

> We never decide for the user 

How do you ask the community then about changes, architecture and roadmap? I have never seen a poll or active polling for community feedback.

Reply:
We should have `fetchOrder` and `fetchMyTrades` and `fetchOrderTrades` properly implemented first (where available) otherwise we won't be able to build more complex functionality (including the caching). We're currently working on the implementation of fundamentals. The next steps towards more unification is something we're going to do when we have the lower level unified to a decent degree (we're close).

> Then I would suggest to introduce a method fetchOrderAndTradesUsingCache to bring my requirements into a function name.

Exactly! And that's what I was trying to explain here as well: https://github.com/ccxt/ccxt/pull/4431#discussion_r245432426. Do you agree to that post btw?

Reply:
> fetchOrderWithFills

Thats the issue with ccxt: it wouldnt be just "order with fills"... without fills binance is not providing the fees for the trade. And god knows what else is missing for other exchanges (similar issue with bitstamp for example)...

`ceateLimitOrder` and `createMarketOrder` already return so different data from each exchange that a call to fetchOrder (not very efficient) is not avoidable. As an app developer I need to check EVERY property if it exist, is filled and the filling makes sense (e.g. if there is a filling and no fees are present but the exchange has fees -> there must be fees so the object is invalid).

Reply:
> I dont think so. I'm doing this already but the fact that I need to implement it for each exchange individually to get all data brings up the question why I'm really using ccxt.

That's again up to the user. Well, yes, you have to code something, there's no library that will code every bit of your logic for you ) So, you might want to use CCXT because it already does some low-level unification that you can build upon, or, alternatively, you might skip CCXT and try recreating those low-level wrappers to build your higher-level caching layers for your own app yourself. I think if you do everything from scratch – that is harder apriori (because you will have to reinvent the wheel) and that is actually a good reason to continue using CCXT and contribute to CCXT, no?

Reply:
> Does that snippet of pseudocode clarify the goals of this library? Because if you're designing the library from bottom upwards – you should agree to the comments in the snippet above )

For me it seems that it is not intended to make progress for app developers and keep it as an (inefficient without using caching) REST wrapper

> The point is that you want it to implement all possible wrappers from the very beginning, but this is not how it works, unfortunately ) If you could make it all complete and ideal from the very start – it would be finished already ) So, we are doing a work in progress that has among one of its goals the goal of designing enough functionality to build the ultimate fetchOrderWrapper that you'd like to see there from the very start )

Well in my experience you can either improve step-by-step (what I wanted to encourage) or start from scratch. I dont think that you can easily put a wrapper on top, then you loose a lot of efficiency (e.g. avoiding calls). I also thought about it but it doesnt make sense for me.  :see_no_evil: 

Dont get me wrong: you did a great job so far but I don't see the advantage for developers/companies that want to benefit from a unifying stable library if there is no change starting in design.

For me unifying means: I can call every exchange the same and get the same unified objects back from all exchanges. You could ask the community if they would like ccxt to become such a library or they prefer to implement workarounds for each exchange individually in their applications.

I wonder why someone is putting so much effort in an "pre-version" library instead (as you called it in the end) instead of doing it right from the beginning. Maybe less exchanges but at least they all work the same. In my experience this pays out in the end (not talking about perfectionism, this usually doesnt pay out). I really would like to contribute for a step-by-step improvement because I can't wait for the perfect-fits-all-wrapper-on-top-of-ccxt-in-the-further-future :-) But then you need to revise your not written, not fixed, design decisions which are done case by case by a small central team without much community involvement (AFAIK).

Reply:
> all your "not written and fixed design decisions" 

Those are actually written here: https://github.com/ccxt/ccxt/blob/master/CONTRIBUTING.md#derived-exchange-classes

![contributing](https://user-images.githubusercontent.com/1294454/50714880-981c4980-108b-11e9-982e-ecf3efad9120.jpg)

> I wonder why someone is putting so much effort in an "pre-version" library instead (as you called it in the end) instead of doing it right from the beginning.

Because we're doing it right from the very beginning. You see?) We just have different views on what is right and what is wrong to do from the very beginning. We believe it is right to reach a unified state by making progress gradually – fundamentals first, more complex things and more data abstraction second.

I'd close this at this point, if you don't mind. We will still be happy to hear from you in the future and to see more contributions from you.

Reply:
> I think if you do everything from scratch – that is harder apriori (because you will have to reinvent the wheel) and that is actually a good reason to continue using CCXT and contribute to CCXT, no?

This depends of the fact if we can agree on a new set of functions that ccxt will officially support ;-) I would like to continue but if I spend days in discussions and rejected commits because of unwritten laws its not fun at all and in the end maybe also not a timer safer ;-)
I would like to start a new set of functions that returns complete objects based on a cache (which is easily to persist) to take advantage of less calls (=efficient library). For example if trades are already present in the cache (no need to call closed orders, but I can call open orders and  compare the filled amount - if present - with the amount in trades -> maybe no call then)

Otherwise a simple fork for the moment would be easier, implementing my logic for myself and everybody that wants to use it. ;-)

Reply:
> I'd close this at this point, if you don't mind. We will still be happy to hear from you in the future and to see more contributions from you.

I would be happy if you can read at least all my comments in here, then it's ok to close :-)

Reply:
> This depends of the fact if we can agree on a new set of functions that ccxt will officially support ;-)

I think we can, but let's make it clear: we only accept code that works universally. If it works for your case but fails on other use cases – we should reconsider it.

> I would like to continue but if I spend days in discussions and rejected commits because of unwritten laws its not fun at all and in the end maybe also not a timer safer ;-)

I'm not interesting in wasting my time either )

> I would like to start a new set of functions that returns complete objects based on a cache (which is easily to persist) to take advantage of less calls (=efficient library). For example if trades are already present in the cache (no need to call closed orders, but I can call open orders and compare the filled amount - if present - with the amount in trades -> maybe no call then)

That works for me – that's exactly what `fetchOrderWithTrades` snippet that I posted above is trying to do. It will be exchange-specific for exchanges that support it, and it will try to workaround the missing fills by calling `fetchOrderTrades` and/or `fetchMyTrades` if the exchange does not provide that info within the order from the very start.

> Otherwise a simple fork for the moment would be easier, implementing my logic for myself and everybody that wants to use it. ;-)

 As long as we're open source – there's tons of forks who think they might really do better )

Reply:
> I would be happy if you can read at least all my comments in here 

I do read all your comments carefully, it just so happens that our replies are out of sync ) But I do read them all, really.

> (especially a reply to the question regarding new methods and if yes which), then I'm fine to close :-)

How about starting literally with the above implementation of `fetchOrderWithTrades`? Do you see how it is doing the same thing that you tried to do in `fetchOrder` for Binance? Do you understand the flaws of this approach when it comes to fills for big orders that hang open for a long while?